STM32: Run time GPIO clocks gating

erwango opened the pull request:
Evolution of #32495 using dedicated device_pm framework

Tested on a blinky sample, with the following settings:
```
CONFIG_PM=y
CONFIG_PM_DEVICE=y
CONFIG_PM_DEVICE_RUNTIME=y
```

Now based on #34781.

Using these settings:
- GPIO port clock is enabled only when needed:
  - When GPIO status is !GPIO_DISCONNECTED and !GPIO_INT_MODE_DISABLED (gpio used by application)
  - and GPIO clock not needed for pinmux configuration
  - In other cases GPIO clock is on and device set as busy
- SoC low power state is entered asynchronously if no device set as busy (in which case we reach 2/3 uA)

EDIT: New PR description after the whole stuff is now actually working.
EDIT: 04-05: Pushed curated version based on latest #34781
EDIT: 04-60: Rebased on master after #34698 was merged. Removed #34781 as not rebased

pabigot commented:
> When adding a device_pm requests to a gpio client (`pinmux_stm32.c`, cf last commit),
> PM framework get stucks when `device_pm_get/put_sync` are called in pinmux_stm32.c
> In that case, `pm_work_handler` is never executed:
> In `device_pm_request` > `k_work_submit` > `k_queue_append(&work_q->queue, work);`
> I end up in `sys_dlist_get`, with a `z_arm_usage_fault` triggered when running `sys_dlist_remove(node);`.

I don't think this is surprising.  It appears the board files invoke `pinmux_stm32_init` which invokes `stm32_setup_pins` at `PRE_KERNEL_1`, which is before it's permitted to invoke kernel API like `k_work_submit()` or `k_queue_append()`.

erwango commented:
> I don't think this is surprising. It appears the board files invoke pinmux_stm32_init which invokes stm32_setup_pins at PRE_KERNEL_1, which is before it's permitted to invoke kernel API like k_work_submit() or k_queue_append().

Thanks for feedback. I though about that as well, but I tried to mode to POST_KERNEL and faced the same issue. So I preferred to stay with PRE_KERNEL_1.
This being said, I can change back to POST_KERNEL and investigate more in that context.

pabigot commented:
I'm not sure what the right path forward is.  It's not clear whether we need to invoke device (or other PM) operations before the kernel gets started or from interrupts (I would think "yes").  If so we have to have a `pre-kernel-ok` path through those functions.  A `no-sleep` path is not sufficient.

erwango commented:
I managed to get things working with the following modifications

- Moving both GPIO and UART to POST_KERNEL
- Disabling BOOT_BANNER
- Replacing `device_pm_enable/device_pm_disable` by `device_pm_get_sync/device_pm_put`

erwango commented:
This requires a bit more work:

Using a modified version of blinky (releasing GPIO device when led is off), I don't enter sleep state.

ceolin commented:
> I managed to get things working with the following modifications
> 
>     * Moving both GPIO and UART to POST_KERNEL
> 
>     * Disabling BOOT_BANNER
> 
>     * Replacing `device_pm_enable/device_pm_disable` by `device_pm_get_sync/device_pm_put`
> 

Do you know why this was necessary ? Because of semaphore usage ?

> 
> Thanks @pabigot to make me investigate more in the good direction.
> 
> > It's not clear whether we need to invoke device (or other PM) operations before the kernel gets started or from interrupts (I would think "yes")
> 
> I'd also say yes.

I think we should be able as well.

erwango commented:
> > I managed to get things working with the following modifications
> > ```
> > * Moving both GPIO and UART to POST_KERNEL
> > 
> > * Disabling BOOT_BANNER
> > 
> > * Replacing `device_pm_enable/device_pm_disable` by `device_pm_get_sync/device_pm_put`
> > ```
> 
> Do you know why this was necessary ? Because of semaphore usage ?

When pm pm_device_get/put primitives are used before kernel init the error is as follows:
PM framework get stucks when device_pm_get/put_sync are used in pinmux_stm32.c
In that case, pm_work_handler is never executed.
In device_pm_request > k_work_submit > k_queue_append(&work_q->queue, work); 
I end up in sys_dlist_get, with a z_arm_usage_fault triggered when running sys_dlist_remove(node);.

erwango commented:
> Using a modified version of blinky (releasing GPIO device when led is off), I don't enter sleep state.

This is now fixed.

erwango committed:
samples/boards: stm32: Add low power blinky sample

This sample demonstrates a low power blinky application.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
@ceolin, @nashif I'm pushing a cleaner version of this PR focused on a blinky use case.
I've omitted button and GPIO interrupts to ease the discussion.

I've added a modified blinky sample that allows to release GPIO resource and enter sleep mode.
This development allows to run `tests/drivers/gpio` tests in the following configurations:
- `CONFIG_PM_DEVICE=y`
- `CONFIG_PM_DEVICE_IDLE=y`

This allows me to raise 3 points:
- Use of `pm_constraint_set/release` vs `device_busy_set/get` in devices. Since we'll eventually have to be able to set constraints in devices, is there a benefit in the use of `device_busy_set/get` ?
- In PM device idle case, I should use `device_pm_get/put` on pinmux side. Though, this is breaking gpio driver behavior in a nonn obvious way. This could be seen when running `tests/drivers/gpio/gpio_basic_api` with `CONFIG_PM_DEVICE_IDLE` (cf FIXME in https://github.com/erwango/zephyr/commit/a8268d08b51dccf8f5c3e78d923d1832975327d9)
- I'm observing a unexpected behavior with `CONFIG_PM_DEVICE_IDLE`. When the system exit sleep mode, it will wake up all devices, starting all the clocks. This is not expected as the clocks should be enabled only if needed. I don't know if this is coming from my implementation or a flaw in PM framework.

![resume_all_devices](https://user-images.githubusercontent.com/18574308/111490990-c271f300-873b-11eb-8bb0-5fcc33b0b36e.png)
At the second blink, all clocks are on...

ceolin commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -561,6 +581,63 @@ static const struct gpio_driver_api gpio_stm32_driver = {
+}
+
+int gpio_stm32_set_power_state(const struct device *dev,
+					      uint32_t new_state)
```
it should be `static` as well.

ceolin commented on samples/boards/stm32/power_mgmt/blinky/src/main.c:
```diff
@@ -0,0 +1,39 @@
+		return;
+	}
+
+	while (1) {
```
nitpick, use `while (true)`  -> https://docs.zephyrproject.org/latest/guides/coding_guidelines/index.html (rule 14.4)

ceolin commented on drivers/gpio/gpio_stm32.h:
```diff
@@ -228,6 +228,10 @@ struct gpio_stm32_data {
 	sys_slist_t cb;
+#ifdef CONFIG_PM_DEVICE
+	/* device power state */
+	uint32_t power_state;
```
that is fine for now IMHO. But we should be able `struct device_pm` field `fms_state`. But I'm looking it right now and it is so confuse that may be better to keep this way right now.

erwango replied:
From what I can see in `pm_work_handler`, these seems like 2 different things.
But this could be reviewed

ceolin commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -577,7 +654,17 @@ static int gpio_stm32_init(const struct device *device)
+	return 0;
+#else
+#ifdef CONFIG_PM_DEVICE
+	data->power_state = DEVICE_PM_ACTIVE_STATE;
```
omg, we also have `DEVICE_PM_STATE_ACTIVE`

erwango replied:
Yeah, it seems that `device->pm->fsm_state` (`DEVICE_PM_STATE_FOO`) and "Device Power Management States" (as used in `device_set_power_state` for instance are 2 different things....

ceolin replied:
yep, got it, just wondering if we should not unify this. Btw, that is the reason I have no complains with the pr, because they are different things right now

ceolin commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -561,6 +581,63 @@ static const struct gpio_driver_api gpio_stm32_driver = {
+
+	}
+
+	if (cb) {
```
`gpio_stm32_set_power_state` is always sync ?

erwango replied:
`gpio_stm32_set_power_state` is always sync, yes. But tbh, I'm not clear on that part. What is the expectation here?

ceolin replied:
> @ceolin Thanks for the review (I'll fix what could be fixed). But I would also have liked to have feedback on the points I reported in the comment [above](https://github.com/zephyrproject-rtos/zephyr/pull/32788#issuecomment-801166195)

Regarding the first point, I think the combination of system constraints and `device_pm_put/get` suppress the need of that API.
The second point is not clear for me what is the problem, if it is just a bug or some architectural problem. Need to debug to understand better.
The third point I'm not sure what we should do, maybe devices should do what they do in the system initialization ? or maybe just restore from what they were before sleep ?

ceolin commented on drivers/pinmux/stm32/pinmux_stm32.c:
```diff
@@ -127,15 +127,35 @@ int stm32_dt_pinctrl_configure(const struct soc_gpio_pinctrl *pinctrl,
+		/* Use direct clock gatting function as a workaround */
+		/* BEWARE that this has serious limitations in case */
+		/* of multithreaded application */
+		/* ret = device_pm_get_sync(port_device); */
```
what happens here ? it seems the key point of having a device pm interface.

erwango replied:
As mentioned in the commit message, using `device_pm_get_sync` here produces strange gpio driver behavior (test/drivers/gpio/gpio_basic_api failing).
So, indeed this should be fixed before merging. I'll try to investigate more on this today.

ceolin commented:
The test is failing on CI. The pr looks good for me, there are things in the device pm that are very weird IMHO but they are orthogonal to this pr.

erwango commented:
@ceolin Thanks for the review (I'll fix what could be fixed). But I would also have liked to have feedback on the points I reported in the comment [above](https://github.com/zephyrproject-rtos/zephyr/pull/32788#issuecomment-801166195)

erwango commented:
@ceolin about this [comment](https://github.com/zephyrproject-rtos/zephyr/pull/32788#discussion_r596447255)
The problem I observe is the following:
> In PM device idle case, I should use device_pm_get/put on pinmux side. Though, this is breaking gpio driver behavior in a nonn obvious way. This could be seen when running tests/drivers/gpio/gpio_basic_api with CONFIG_PM_DEVICE_IDLE (cf FIXME in erwango@a8268d0)

After a debugging session, what I can observe is the following:
- In one particular call to `gpio_stm32_config`, when calling `device_pm_get_sync(dev)`, `pm_work_handler` is not executed. This leads to a GPIO clock remaining off while it is expected to be on and eventually to a gpio behavior issue and test is failed.
- If I'm not calling `device_pm_get_sync(dev)` in pinmux function (`stm32_dt_pinctrl_configure`, hence the FIXME line), the same call to  `device_pm_get_sync(dev)` in function `gpio_stm32_config` leads to `pm_work_handler` execution as expected, and test is passed.

ceolin commented on samples/boards/stm32/power_mgmt/blinky/src/main.c:
```diff
@@ -0,0 +1,39 @@
+		gpio_pin_set(led.port, led.pin, (int)led_is_on);
+		if (led_is_on == false) {
+			/* Release resource to allow sleep mode */
+			gpio_pin_configure(led.port, led.pin, GPIO_DISCONNECTED);
```
shouldn't it be `gpio_pin_configure_dt(&led, GPIO_DISCONNECTED)` or `gpio_pin_configure(led.port, led.pin, led.dt_flags | GPIO_DISCONNECTED)` ?

erwango replied:
As mentioned, I've implemented a quite binary check on driver side to identify if port is in  use or not.
If `flag != GPIO_DISCONNECTED`, I consider it is busy.
So configuring dt flags will prevent the system to enter sleep states.

pabigot replied:
That seems fair, though unless there's some value in having pull and drain flags have an effect when the GPIO is disconnected (which would inevitably be driver-specific, as some hardware wouldn't respect that) the driver could just as well determine that if a GPIO is neither an input nor an output it's not in use, regardless of other flags.  I believe that's consistent with the discussion in #32220.

erwango replied:
@pabigot Indeed. Aim of this first version was rather to check the dynamic behavior of the implementation vs PM framework.
I'm working in parallel on a finer grain approach.

erwango commented:
New iteration, focused on simplifying the gpio port clock handling (see commit description for details)

Versus the last push that reported 3 issues (cf https://github.com/zephyrproject-rtos/zephyr/pull/32788#issuecomment-801166195), 2 issues are somehow fixed:
- `pm_constraint` are removed after observation that gpio don't generate background processing. `pm_constraint` should be handled at higher level component (application)
- Non used gpio ports are released systematically after they are being woken up.

Remains the strange behavior observed when using `device_pm_get_sync` in pinmux_stm32.c, getting gpio tests failed.
I've added a commit to ease reproduction in tests.

erwango committed:
drivers/gpio: stm32: Enable PM_DEVICE services

Implement power mgmt hooks to support PM_DEVICE and
PM_DEVICE_RUNTIME.
In case of PM_DEVICE_RUNTIME, clock is requested for bank writes
so it is requested before configuring and released only if pin
is not configured as output.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/pinmux: stm32: Enable PM_DEVICE_RUNTIME services

Add PM_DEVICE_RUNTIME support

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
tests/drivers: gpio: Add support for stm32l562e_dk

Required for board testing


Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
Last push with following includes:
- L5 Support (includes and depends on https://github.com/zephyrproject-rtos/zephyr/pull/33131)
- GPIO PM handling simplifications
  - No more constraint_set/device_busy_set. It is up to application to configure GPIO resource depending on its needs
  - GPIO clock is released when configuration is `!GPIO_OUTPUT`

With this config we achieve following consumption for blinky
![l5_blinky_5uA](https://user-images.githubusercontent.com/18574308/112662968-94d22b80-8e58-11eb-9f1d-6ad7fb005f96.png)
- GPIO clock is released as soon as no more required (once pin set is done)
- Core goes to sleep state (STOP2) as soon as possible (k_sleep() called by application)

Same consumption is seen for both PM_DEVICE and PM_DEVICE_IDLE.

ceolin commented:
> Last push with following includes:
> 
> * L5 Support (includes and depends on #33131)
> * GPIO PM handling simplifications
>   
>   * No more constraint_set/device_busy_set. It is up to application to configure GPIO resource depending on its needs
>   * GPIO clock is released when configuration is `!GPIO_OUTPUT`
> 
> With this config we achieve following consumption for blinky
> ![l5_blinky_5uA](https://user-images.githubusercontent.com/18574308/112662968-94d22b80-8e58-11eb-9f1d-6ad7fb005f96.png)
> 
> * GPIO clock is released as soon as no more required (once pin set is done)
> * Core goes to sleep state (STOP2) as soon as possible (k_sleep() called by application)
> 
> Same consumption is seen for both PM_DEVICE and PM_DEVICE_IDLE.

![power-meter](https://user-images.githubusercontent.com/882919/113201020-e6decc80-921d-11eb-8517-797edc46c803.png)

Are you using the blinky sample right ? any idea why I'm getting different values

erwango commented:
> Are you using the blinky sample right ? any idea why I'm getting different values

As discussed, to achieve low power states it is required to disable CONFIG_DEBUG.

ceolin commented:
> > Are you using the blinky sample right ? any idea why I'm getting different values
> 
> As discussed, to achieve low power states it is required to disable CONFIG_DEBUG.

Yep, that did the trick. Btw take a look in this initial pr https://github.com/zephyrproject-rtos/zephyr/pull/33942

erwango committed:
drivers/gpio: stm32: Don't perform clock gating with VddIO2

In some cases, VddIO2 is required to get port working.
Looking in details, VddIO2 should be set on start up
but not toggled on/off in PM use cases.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
Converted to PR a bit early as strange behavior is still visible with combination of PM and PM_DEVICE_IDLE.
Working closely with @ceolin to fix this.

ABOSTM commented on samples/boards/stm32/power_mgmt/blinky/README.rst:
```diff
@@ -0,0 +1,42 @@
+
+   Unsupported board: led0 devicetree alias is not defined
+
+The board must have an LED connected via a GPIO pin. These are called "User
```
"a LED"

erwango replied:
I think we say "ell ee dee", so "an ell ee dee" would be correct

pabigot replied:
Yes, `an LED` is correct.  https://blog.apastyle.org/apastyle/2012/04/using-a-or-an-with-acronyms-and-abbreviations.html

ABOSTM replied:
Ok, Sorry.

ABOSTM commented on samples/boards/stm32/power_mgmt/blinky/README.rst:
```diff
@@ -0,0 +1,42 @@
+Overview
+********
+
+Blinky is a simple application which blinks an LED forever using the :ref:`GPIO
```
"a LED"

ABOSTM reviewed (commented):
Minor typo, otherwise LGTM.

erwango commented:
Rebased on master after #34698 was merged. Removed #34781 as not rebased

FRASTM commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -484,6 +491,14 @@ static int gpio_stm32_config(const struct device *dev,
+	/* Device released */
+#ifdef CONFIG_PM_DEVICE_RUNTIME
+	/* Release clock only if configuration requires bank writes */
+	if (((flags & GPIO_OUTPUT) != 0)) {
```
`((flags & GPIO_OUTPUT) != 0)`

ABOSTM reviewed (approved):
LGTM

FRASTM commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -561,6 +576,65 @@ static const struct gpio_driver_api gpio_stm32_driver = {
+	struct gpio_stm32_data *data = dev->data;
+	int ret = 0;
+
+	if (new_state == PM_DEVICE_ACTIVE_STATE) {
```
where  `PM_DEVICE_ACTIVE_STATE`  differs from `DEVICE_PM_ACTIVE_STATE` (as defined by the include/device.h) ?

gmarull replied:
@FRASTM all `DEVICE_PM` macros have been moved to the `PM_DEVICE` namespace

erwango replied:
PM_DEVICE_ACTIVE_STATE is the new DEVICE_PM_ACTIVE_STATE in latest master.
Change occured in #34698

FRASTM replied:
ok

gmarull commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -474,6 +475,12 @@ static int gpio_stm32_config(const struct device *dev,
 	}
 
+
+#ifdef CONFIG_PM_DEVICE_RUNTIME
+	/* Enable device clock before configuration (requires bank writes) */
+	pm_device_get_sync(dev);
+#endif /* CONFIG_PM_DEVICE */
```
note: sample uses return instead of `goto exit`. As this function is not doing any sort of cleanup goto should be replaced by just return.

gmarull commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -484,6 +491,14 @@ static int gpio_stm32_config(const struct device *dev,
+#ifdef CONFIG_PM_DEVICE_RUNTIME
+	/* Release clock only if configuration requires bank writes */
+	if (((flags & GPIO_OUTPUT) != 0)) {
+		pm_device_put(dev);
```
error checking, see my previous comment

gmarull commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -474,6 +475,15 @@ static int gpio_stm32_config(const struct device *dev,
+#ifdef CONFIG_PM_DEVICE_RUNTIME
+	/* Enable device clock before configuration (requires bank writes) */
+	err = pm_device_get_sync(dev);
+	if (err < 0 && err != -ENOSYS) {
```
when checking for `-ENOSYS` conditional compilation is not needed, since it will never happen.

erwango replied:
Right. So actually, I'll keep the CONFIG_PM_DEVICE_RUNTIME, as I need to do some more stuff at this point.

FRASTM commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -250,9 +251,9 @@ int gpio_stm32_clock_request(const struct device *dev, bool on)
 
 #if defined(PWR_CR2_IOSV) && DT_NODE_HAS_STATUS(DT_NODELABEL(gpiog), okay)
 	z_stm32_hsem_lock(CFG_HW_RCC_SEMID, HSEM_LOCK_DEFAULT_RETRY);
+	/* Port G[15:2] requires external power supply */
```
this modification could also come in the commit  "drivers/gpio: stm32: Don't perform clock gating with VddIO2 " 38c48f3af5bc47a2d2c31bdce731508677106951

erwango replied:
Fixed

erwango commented on samples/boards/stm32/power_mgmt/blinky/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+
+cmake_minimum_required(VERSION 3.13.1)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(stm32_lp_blinky)
```
done

erwango commented on samples/boards/stm32/power_mgmt/blinky/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
```
Thanks. I'll keep it that way for now.
But open to change after.

gmarull commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -697,7 +697,7 @@ static int gpio_stm32_init(const struct device *dev)
 			    &gpio_stm32_data_## __suffix,		       \
 			    &gpio_stm32_cfg_## __suffix,		       \
-			    POST_KERNEL,				       \
+			    PRE_KERNEL_1,				       \
```
Is this still needed?

erwango replied:
Nope, txs

gmarull commented on drivers/gpio/gpio_stm32.c:
```diff
@@ -561,6 +572,65 @@ static const struct gpio_driver_api gpio_stm32_driver = {
+		ret = gpio_stm32_clock_request(dev, false);
+	}
+
+	data->power_state = new_state;
```
Shouldn't you return if ret < 0 and not change the state?
